src/transforms/transaction.rs

```diff
@@ -0,0 +1,812 @@
+                            }
+                            Err(err) => {
+                                warn!("failed to merge value: {}", err);
+                            }
+                        }
+                    } else {
+                        entry.insert(v.clone().into());
+                    }
+                }
+                hash_map::Entry::Occupied(mut entry) => {
+                    if let Err(err) = entry.get_mut().add(v.clone()) {
+                        warn!("failed to merge value: {}", err);
+                    }
+                }
+            }
+        }
+        self.stale_since = Instant::now();
+    }
+
+    fn flush(mut self) -> LogEvent {
+        let mut event = Event::new_empty_log().into_log();
+        for (k, v) in self.fields.drain() {
+            if let Err(err) = v.insert_into(k, &mut event) {
+                warn!("failed to merge values for field: {}", err);
+            }
+        }
+        event
+    }
+}
+
+//------------------------------------------------------------------------------
+
+pub struct Transaction {
+    expire_after: Duration,
+    flush_period: Duration,
+    identifier_fields: Vec<Atom>,
+    merge_strategies: IndexMap<String, MergeStrategy>,
+    transaction_merge_states: HashMap<Discriminant, TransactionState>,
+    ends_when: Option<Box<dyn Condition>>,
+}
+
+impl Transaction {
+    fn new(config: &TransactionConfig) -> crate::Result<Self> {
+        let ends_when = if let Some(ends_conf) = &config.ends_when {
+            Some(ends_conf.build()?)
+        } else {
+            None
+        };
+
+        let identifier_fields = config
+            .identifier_fields
+            .clone()
+            .into_iter()
+            .map(Atom::from)
+            .collect();
+
+        Ok(Transaction {
+            expire_after: Duration::from_millis(config.expire_after_ms.unwrap_or(30000)),
+            flush_period: Duration::from_millis(config.flush_period_ms.unwrap_or(1000)),
+            identifier_fields,
+            merge_strategies: config.merge_strategies.clone(),
+            transaction_merge_states: HashMap::new(),
+            ends_when,
+        })
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) {
+        let mut flush_discriminants = Vec::new();
+        for (k, t) in &self.transaction_merge_states {
+            if t.stale_since.elapsed() >= self.expire_after {
+                flush_discriminants.push(k.clone());
+            }
+        }
+        for k in &flush_discriminants {
+            if let Some(t) = self.transaction_merge_states.remove(k) {
+                output.push(Event::from(t.flush()));
+            }
+        }
+    }
+
+    fn flush_all_into(&mut self, output: &mut Vec<Event>) {
+        self.transaction_merge_states
+            .drain()
+            .for_each(|(_, s)| output.push(Event::from(s.flush())));
+    }
+}
+
+impl Transform for Transaction {
+    // Only used in tests
```

Comment:
Can it get `#[cfg(test)]`?

Reply:
Sadly not with the current `Transform` trait, but that'll be fixed with #2939.